File: src/types/_ActivatableSkill.ts
/**
 * @title Activatable Skill
 */

import * as DB from "tsondb/schema/dsl"
import { FastCastingTime, SlowCastingTime } from "./_ActivatableSkillCastingTime.js"
import { OneTimeCost, SustainedCost } from "./_ActivatableSkillCost.js"
import { DurationForOneTime, DurationForSustained } from "./_ActivatableSkillDuration.js"
import { Range } from "./_ActivatableSkillRange.js"

export const OneTimePerformanceParameters = DB.GenTypeAlias(import.meta.url, {
  name: "OneTimePerformanceParameters",
  parameters: [DB.Param("CastingTime")],
  type: CastingTime =>
    DB.Object({
      casting_time: DB.Required({
        type: DB.TypeArgument(CastingTime),
      }),
      cost: DB.Required({
        type: DB.IncludeIdentifier(OneTimeCost),
      }),
      range: DB.Required({
        type: DB.IncludeIdentifier(Range),
      }),
      duration: DB.Required({
        type: DB.IncludeIdentifier(DurationForOneTime),
      }),
    }),
})

const SustainedPerformanceParameters = DB.GenTypeAlias(import.meta.url, {
  name: "SustainedPerformanceParameters",
  parameters: [DB.Param("CastingTime")],
  type: CastingTime =>
    DB.Object({
      casting_time: DB.Required({
        type: DB.TypeArgument(CastingTime),
      }),
      cost: DB.Required({
        type: DB.IncludeIdentifier(SustainedCost),
      }),
      range: DB.Required({
        type: DB.IncludeIdentifier(Range),
      }),
      duration: DB.Optional({
        type: DB.IncludeIdentifier(DurationForSustained),
      }),
    }),
})

export const FastPerformanceParameters = DB.Enum(import.meta.url, {
  name: "FastPerformanceParameters",
  values: () => ({
    OneTime: DB.EnumCase({ type: DB.IncludeIdentifier(FastOneTimePerformanceParameters) }),
    Sustained: DB.EnumCase({ type: DB.IncludeIdentifier(FastSustainedPerformanceParameters) }),
  }),
})

const FastOneTimePerformanceParameters = DB.TypeAlias(import.meta.url, {
  name: "FastOneTimePerformanceParameters",
  type: () =>
    DB.GenIncludeIdentifier(OneTimePerformanceParameters, [DB.IncludeIdentifier(FastCastingTime)]),
})

const FastSustainedPerformanceParameters = DB.TypeAlias(import.meta.url, {
  name: "FastSustainedPerformanceParameters",
  type: () =>
    DB.GenIncludeIdentifier(SustainedPerformanceParameters, [
      DB.IncludeIdentifier(FastCastingTime),
    ]),
})

export const SlowPerformanceParameters = DB.Enum(import.meta.url, {
  name: "SlowPerformanceParameters",
  values: () => ({
    OneTime: DB.EnumCase({ type: DB.IncludeIdentifier(SlowOneTimePerformanceParameters) }),
    Sustained: DB.EnumCase({ type: DB.IncludeIdentifier(SlowSustainedPerformanceParameters) }),
  }),
})

const SlowOneTimePerformanceParameters = DB.TypeAlias(import.meta.url, {
  name: "SlowOneTimePerformanceParameters",
  type: () =>
    DB.GenIncludeIdentifier(OneTimePerformanceParameters, [DB.IncludeIdentifier(SlowCastingTime)]),
})

const SlowSustainedPerformanceParameters = DB.TypeAlias(import.meta.url, {
  name: "SlowSustainedPerformanceParameters",
  type: () =>
    DB.GenIncludeIdentifier(SustainedPerformanceParameters, [
      DB.IncludeIdentifier(SlowCastingTime),
    ]),
})

export const OldParameter = DB.TypeAlias(import.meta.url, {
  name: "OldParameter",
  type: () =>
    DB.Object({
      full: DB.Required({
        type: DB.String(),
      }),
      abbr: DB.Required({
        type: DB.String(),
      }),
    }),
})

File: Sources/OptolithDatabaseSchema/GeneratedTypes/_ActivatableSkill.swift
//
//  _ActivatableSkill.swift
//  OptolithDatabaseSchema
//

public struct OneTimePerformanceParameters<CastingTime: EntitySubtype>: EntitySubtype {
    public let castingTime: CastingTime

    public let cost: OneTimeCost

    public let range: Range

    public let duration: DurationForOneTime

    public init(castingTime: CastingTime, cost: OneTimeCost, range: Range, duration: DurationForOneTime) {
        self.castingTime = castingTime
        self.cost = cost
        self.range = range
        self.duration = duration
    }

    private enum CodingKeys: String, CodingKey {
        case castingTime = "casting_time"
        case cost = "cost"
        case range = "range"
        case duration = "duration"
    }
}

public struct SustainedPerformanceParameters<CastingTime: EntitySubtype>: EntitySubtype {
    public let castingTime: CastingTime

    public let cost: SustainedCost

    public let range: Range

    public let duration: DurationForSustained?

    public init(castingTime: CastingTime, cost: SustainedCost, range: Range, duration: DurationForSustained? = nil) {
        self.castingTime = castingTime
        self.cost = cost
        self.range = range
        self.duration = duration
    }

    private enum CodingKeys: String, CodingKey {
        case castingTime = "casting_time"
        case cost = "cost"
        case range = "range"
        case duration = "duration"
    }
}

public enum FastPerformanceParameters: EntitySubtype {
    case oneTime(FastOneTimePerformanceParameters)
    case sustained(FastSustainedPerformanceParameters)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case oneTime = "one_time"
        case sustained = "sustained"
    }

    private enum Discriminator: String, Decodable {
        case oneTime = "OneTime"
        case sustained = "Sustained"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .oneTime:
            self = .oneTime(try container.decode(FastOneTimePerformanceParameters.self, forKey: .oneTime))
        case .sustained:
            self = .sustained(try container.decode(FastSustainedPerformanceParameters.self, forKey: .sustained))
        }
    }
}

public typealias FastOneTimePerformanceParameters = OneTimePerformanceParameters<FastCastingTime>

public typealias FastSustainedPerformanceParameters = SustainedPerformanceParameters<FastCastingTime>

public enum SlowPerformanceParameters: EntitySubtype {
    case oneTime(SlowOneTimePerformanceParameters)
    case sustained(SlowSustainedPerformanceParameters)

    private enum CodingKeys: String, CodingKey {
        case tag = "tag"
        case oneTime = "one_time"
        case sustained = "sustained"
    }

    private enum Discriminator: String, Decodable {
        case oneTime = "OneTime"
        case sustained = "Sustained"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let tag = try container.decode(Discriminator.self, forKey: .tag)
        switch tag {
        case .oneTime:
            self = .oneTime(try container.decode(SlowOneTimePerformanceParameters.self, forKey: .oneTime))
        case .sustained:
            self = .sustained(try container.decode(SlowSustainedPerformanceParameters.self, forKey: .sustained))
        }
    }
}

public typealias SlowOneTimePerformanceParameters = OneTimePerformanceParameters<SlowCastingTime>

public typealias SlowSustainedPerformanceParameters = SustainedPerformanceParameters<SlowCastingTime>

public struct OldParameter: EntitySubtype {
    public let full: String

    public let abbr: String

    public init(full: String, abbr: String) {
        self.full = full
        self.abbr = abbr
    }
}
